core/src/main/java/org/elasticsearch/action/index/IndexResponse.java

```diff
@@ -67,4 +78,27 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         builder.field("created", result == Result.CREATED);
         return builder;
     }
+
+    private static final ConstructingObjectParser<IndexResponse, ParseFieldMatcherSupplier> PARSER;
+    static {
+        PARSER = new ConstructingObjectParser<>(IndexResponse.class.getName(),
+                args -> {
+                    // index uuid and shard id  are unknown
+                    ShardId shardId = new ShardId(new Index((String) args[0], IndexMetaData.INDEX_UUID_NA_VALUE), -1);
+                    String type = (String) args[1];
+                    String id = (String) args[2];
+                    long version = (long) args[3];
+                    Result result = Result.parse((String) args[4]);
+                    long seqNo = (args[5] != null) ? (long) args[5] : -1;
+                    boolean created = (boolean) args[6];
+                    assert created == (result == Result.CREATED);
+                    return new IndexResponse(shardId, type, id, seqNo, version, created);
+                });
+        DocWriteResponse.declareParserFields(PARSER);
```

Comment:
is it correct to do this in IndexResponse? I am a bit confused by who does what. Given the toXContent in IndexResponse that only prints out one field, I was expecting most of the parsing to be in DocWriteResponse. That may have to do again with me not understanding ObjectParser and how it works.